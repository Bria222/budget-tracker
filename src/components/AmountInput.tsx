import * as React from "react";
import { TextInput } from "./TextInput";

interface AmountInputProps {
    label?: string;
    amountInput?: number;
    onAmountChange: (amount: number) => void;
    helperText?: string;
}

export const AmountInput: React.FC<AmountInputProps> = (props) => {

    const handleAmountChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const amountFloat = parseFloat(event.target.value);
        props.onAmountChange(amountFloat);
    }

    return (
        <TextInput 
            autoFocus
            required
            type='number'
            label={props.label || 'Amount'}
            value={props.amountInput || ''}
            inputProps={{ step: '.01', 'aria-required': true }}
            onChange={handleAmountChange}
            helperText={props.helperText}
        />
    );           
}

