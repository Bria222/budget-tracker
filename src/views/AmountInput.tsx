import * as React from "react";
import CircularProgress from '@material-ui/core/CircularProgress';
import Grid from "@material-ui/core/Grid";
import { CurrencyInput, CurrencyInputProps } from "./CurrencyInput";
import { TextInput } from "./TextInput";
import { Typography } from "@material-ui/core";
import { currenciesStore } from "../stores/CurrenciesStore";


interface AmountInputProps {
    label?: string;
    amount?: number;
    onAmountChange?: (amount: number) => void;
    helperText?: string;
}

interface AmountInputState {
    amount: string;
}

export class AmountInput extends React.PureComponent<AmountInputProps, AmountInputState> {

    constructor(props: AmountInputProps) {
        super(props);
        this.state = {amount: props.amount ? `${props.amount}` : ''};
    }

    private handleAmountChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({amount: event.target.value});
        if (this.props.onAmountChange) {
            this.props.onAmountChange(parseFloat(event.target.value));
        }
    }
    
    render() {
        if (this.state) {
            return (
                <TextInput 
                    autoFocus
                    required
                    type='number'
                    label={this.props.label || 'Amount'}
                    value={this.state.amount}
                    inputProps={{ min: '0', step: '1', 'aria-required': true }}
                    onChange={this.handleAmountChange}
                    helperText={this.props.helperText}
                />
            );
        }
        return <CircularProgress />;
    }           
}

interface ExpenseAmountInputProps extends CurrencyInputProps, AmountInputProps {
    baseCurrency: string;
}

interface ExpenseAmountInputState {
    amountBaseCurrency?: number;
}

export class AmountWithCurrencyInput extends React.PureComponent<ExpenseAmountInputProps, ExpenseAmountInputState> {

    constructor(props: ExpenseAmountInputProps) {
        super(props);
        this.state = { };
    }

    render () {
        return (
            <Grid container direction='row' alignItems='baseline'>
                <Grid item>
                    <AmountInput {...this.props} onAmountChange={this.handleAmountChange} helperText={this.baseAmount}/>
                </Grid>
                <Grid item >
                    <CurrencyInput {...this.props} onCurrencyChange={this.handleCurrencyChange}/>
                </Grid>
            </Grid>);
    }

    get baseAmount () {
        if (this.isDifferentCurrency) {
            return `${this.state.amountBaseCurrency} ${this.props.baseCurrency}`;
        }
    }

    handleAmountChange = (amount: number) => {
        if (this.props.onAmountChange) {
            this.props.onAmountChange(amount);
        }
        if (this.isDifferentCurrency) {
            this.calculateBaseAmount();
        }
    }

    handleCurrencyChange = (currency: string) => {
        if (this.props.onCurrencyChange) {
            this.props.onCurrencyChange(currency);
        }
        if (this.isDifferentCurrency) {
            this.calculateBaseAmount();
        }
    }

    private async calculateBaseAmount () {
        const rate = await currenciesStore.getRate(
            this.props.baseCurrency, 
            this.props.selectedCurrency);
        if (rate) {
            this.setState({amountBaseCurrency: Math.round(this.props.amount *100 / rate) / 100 });
        }
    }

    get isDifferentCurrency () {
        return this.props.baseCurrency !== this.props.selectedCurrency;
    }
}
