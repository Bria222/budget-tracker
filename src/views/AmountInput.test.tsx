import React from 'react';
import renderer from 'react-test-renderer';

import { AmountInput } from './AmountInput';

test('AmountInput value is 11', () => {
  const tree = renderer
    .create(<AmountInput amountInput={11} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
