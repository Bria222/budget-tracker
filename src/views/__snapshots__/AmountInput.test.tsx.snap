// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput value is 11 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root"
  style={
    Object {
      "margin": 8,
    }
  }
>
  <label
    className="MuiFormLabel-root MuiFormLabel-filled Mui-required Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense"
    data-shrink={true}
    htmlFor="input-field-Amount"
  >
    Amount
    <span
      className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
    >
      â€‰
      *
    </span>
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      aria-required={true}
      autoFocus={true}
      className="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
      disabled={false}
      id="input-field-Amount"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      step=".01"
      type="number"
      value={11}
    />
  </div>
</div>
`;
